types/underscore/index.d.ts

```diff
@@ -1038,19 +1025,19 @@ declare module _ {
             context?: any): number;
 
         /**
-        * Uses a binary search to determine the index at which the value should be inserted into the list in order
-        * to maintain the list's sorted order. If an iterator is passed, it will be used to compute the sort ranking
-        * of each value, including the value you pass.
-        * @param list The sorted list.
-        * @param value The value to determine its index within `list`.
-        * @param iterator Iterator to compute the sort ranking of each value, optional.
-        * @param context `this` object in `iterator`, optional.
-        * @return The index where `value` should be inserted into `list`.
-        **/
-        sortedIndex<T, TSort>(
-            list: _.List<T>,
-            value: T,
-            iterator?: ((x: T) => TSort) | string,
+         * Uses a binary search to determine the index at which the value should be inserted into `list` in order
+         * to maintain `list`'s sorted order. If an iteratee is provided, it will be used to compute the sort ranking
+         * of each value, including the value you pass.
+         * @param list A sorted list.
+         * @param value The value to determine an insert index for to mainain the sorting in `list`.
+         * @param iteratee Iteratee to compute the sort ranking of each element including `value`, optional.
+         * @param context `this` object in `iteratee`, optional.
+         * @return The index where `value` should be inserted into `list`.
+         **/
+        sortedIndex<V extends List<any>>(
+            list: V,
+            value: TypeOfList<V>,
+            iteratee?: Iteratee<V, any>,
```

Comment:
The matcher-type iteratee doesn't seem super useful for `sortedIndex` or `uniq` (though I guess I could see the possibility of occasional usefulness for it in the latter), so I didn't add tests for them. I'm happy to add such tests if folks object, though.

If there are no objections, would it be worth going a step further by splitting `Iteratee` into two types like `Iteratee = <everything but matcher>` and `MatcherIteratee = Iteratee | Partial<T>` and only using the latter in places where using a matcher seems to make sense?

Reply:
A matcher makes sense even for `sortedIndex`: you would use it if you want to keep all elements that pass the matcher at the end of the array. Or at the beginning, if you `_.negate` it.

Like I prefer keeping `_Chain` a single interface because this most accurately aligns with reality, I'd prefer keeping `Iteratee` and `IterateeResult` single types, too.

Reply:
Per your approval comment, I'll add these tests back in.